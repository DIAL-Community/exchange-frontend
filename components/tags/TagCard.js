import { useIntl } from 'react-intl'

const TagCard = ({ tag, listType }) => {
  const { formatMessage } = useIntl()
  const format = (id, values) => formatMessage({ id: id }, values)

  return (
    <>
      {listType === 'list'
        ? (
          <div className='border-3 border-transparent hover:border-dial-yellow text-button-gray hover:text-dial-yellow cursor-pointer'>
            <div className='border border-dial-gray hover:border-transparent card-drop-shadow'>
              <div className='flex justify-between my-5 px-4'>
                <div className='inline-block card-title card-link-text text-button-gray'>
                  {tag}
                </div>
              </div>
            </div>
          </div>
        )
        : <div>{format('tag.label')}</div>}
    </>
  )
}

export default TagCard
