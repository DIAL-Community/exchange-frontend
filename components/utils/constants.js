export const DisplayType = {
  LARGE_CARD: 'large-card',
  SMALL_CARD: 'small-card',
  PINNED_CARD: 'pinned-card',
  FEATURED_CARD: 'featured-card',
  HUB_CARD: 'hub-card',
  GRID_CARD: 'grid-card'
}

export const MainDisplayType = {
  LIST: 'list',
  GRID: 'grid'
}

export const ORIGIN_SLUG_EXPANSIONS = {
  'dial': 'Digital Impact Alliance',
  'digital-government-platform-tracker': 'Digital Government Platform Tracker',
  'digital-health-atlas': 'Digital Health Atlas',
  'digital-square': 'Digital Square',
  'dpga': 'Digital Public Goods Alliance',
  'giz': 'GIZ',
  'indiastack': 'India Stack',
  'manually-entered': 'Manually Entered',
  'unicef-covid': 'UNICEF'
}

export const OpportunityStatus = {
  CLOSED: 'CLOSED',
  OPEN: 'OPEN',
  UPCOMING: 'UPCOMING'
}

export const OpportunityType = {
  BID: 'BID',
  BUILDING_BLOCK: 'BUILDING BLOCK',
  INNOVATION: 'INNOVATION',
  TENDER: 'TENDER',
  OTHER: 'OTHER'
}

export const MaturityStatus = {
  DRAFT: 'DRAFT',
  PUBLISHED: 'PUBLISHED'
}

export const CategoryType = {
  DPI: 'DPI',
  FUNCTIONAL: 'FUNCTIONAL'
}

export const MappingStatus = {
  BETA: 'BETA',
  MATURE: 'MATURE',
  SELF_REPORTED: 'SELF-REPORTED',
  VALIDATED: 'VALIDATED'
}

export const LicenseTypeFilter = {
  ALL: 'all_license',
  COMMERCIAL: 'commercial_only',
  OPEN_SOURCE:'oss_only'
}

export const ObjectType = {
  BUILDING_BLOCK: 'BUILDING_BLOCK',
  DATASET: 'OPEN_DATA',
  PRODUCT: 'PRODUCT',
  PROJECT: 'PROJECT',
  OPPORTUNITY: 'OPPORTUNITY',
  ORGANIZATION: 'ORGANIZATION',
  USE_CASE: 'USE_CASE',
  WORKFLOW: 'WORKFLOW',
  MOVE: 'MOVE',
  PLAY: 'PLAY',
  PLAYBOOK: 'PLAYBOOK',
  RUBRIC_CATEGORY: 'RUBRIC_CATEGORY',
  CATEGORY_INDICATOR: 'CATEGORY_INDICATOR',
  CANDIDATE_DATASET: 'CANDIDATE_OPEN_DATA',
  CANDIDATE_ORGANIZATION: 'CANDIDATE_ORGANIZATION',
  CANDIDATE_PRODUCT: 'CANDIDATE_PRODUCT',
  CANDIDATE_RESOURCE: 'CANDIDATE_RESOURCE',
  CANDIDATE_ROLE: 'CANDIDATE_ROLE',
  TAG: 'TAG',
  SECTOR: 'SECTOR',
  URL: 'URL',
  CITY: 'CITY',
  COUNTRY: 'COUNTRY',
  REGION: 'REGION',
  TASK: 'TASK',
  USER: 'USER',
  CONTACT: 'CONTACT',
  RESOURCE: 'RESOURCE',
  RESOURCE_TOPIC: 'RESOURCE_TOPIC',
  SITE_SETTING: 'SITE_SETTING',
  TENANT_SETTING: 'TENANT_SETTING'

}

export const CategoryIndicatorType = {
  NUMERIC: 'numeric',
  SCALE: 'scale',
  BOOLEAN: 'boolean'
}

export const CandidateActionType = {
  APPROVE: 'APPROVE',
  REJECT: 'REJECT'
}

export const DEFAULT_PAGE_SIZE = 8

export const ProductStageType = {
  PILOT: 'pilot',
  SCALING: 'scaling',
  MATURE: 'mature'
}

export const ProductExtraAttributeNames = [
  'Relevance',
  'Impact',
  'Local Ownership',
  'Funders'
]

export const ProductMaturityAttributeNames = [
  'Deployments',
  'Deployed Countries',
  'Active Users',
  'Transactions per month',
  'Annual Revenue',
  'Funding Raised'
]
