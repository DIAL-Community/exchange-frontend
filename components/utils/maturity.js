export const getCategoryIndicatorTypes = (format) => [
  { label: format('shared.categoryIndicator.type.numeric'), value: 'numeric' },
  { label: format('shared.categoryIndicator.type.scale'), value: 'scale' },
  { label: format('shared.categoryIndicator.type.boolean'), value: 'boolean' }
]

export const getCategoryIndicatorScaleOptions = (format) => [
  { label: format('shared.categoryIndicator.type.scale.low'), value: 'low' },
  { label: format('shared.categoryIndicator.type.scale.medium'), value: 'medium' },
  { label: format('shared.categoryIndicator.type.scale.high'), value: 'high' }
]

export const getCategoryIndicatorBooleanOptions = (format) => [
  { label: format('shared.categoryIndicator.type.boolean.true'), value: 't' },
  { label: format('shared.categoryIndicator.type.boolean.false'), value: 'f' }
]

export const getCategoryIndicatorNumericOptions = () =>
  ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10'].map(value => ({ label: value, value }))
