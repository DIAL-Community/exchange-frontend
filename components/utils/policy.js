import { GRAPH_QUERY_CONTEXT_KEY, GRAPH_QUERY_POLICY_SLUG } from '../../lib/apolloClient'

const operationAllowed = (error, operationName) => {
  if (error?.graphQLErrors) {
    const { graphQLErrors } = error
    if (graphQLErrors.length > 0) {
      graphQLErrors.forEach(e => {
        if (e.extensions.operation === operationName) {
          return true
        }
      })
    }
  }

  return false
}

// Return map of operation and boolean value for each operation
// e.g. { editing: true, deleting: true }
export const fetchOperationPolicies = async (client, query, operations, variables = { slug: GRAPH_QUERY_POLICY_SLUG }) => {
  const response = await client.query({
    query,
    variables,
    errorPolicy: 'all',
    context: {
      headers: {
        [GRAPH_QUERY_CONTEXT_KEY]: operations.join(', ')
      }
    }
  })

  const policies = {}
  operations.forEach(operation => {
    policies[operation] = operationAllowed(response.error, operation)
  })

  return policies
}
