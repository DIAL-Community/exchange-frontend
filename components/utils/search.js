import { GRAPH_QUERY_CONTEXT } from '../../lib/apolloClient'

const MIN_SEARCH_VALUE_CHARACTERS = 2

export const fetchSelectOptions = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({
    query,
    variables,
    context: {
      headers: {
        ...GRAPH_QUERY_CONTEXT.VIEWING
      }
    }
  })

  return response.data ? fetchedDataCallback(response.data) : []
}

export const fetchSelectOptionsWithMaturity = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    mature: true,
    search: input
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({
    query,
    variables,
    context: {
      headers: {
        ...GRAPH_QUERY_CONTEXT.VIEWING
      }
    }
  })

  return response.data ? fetchedDataCallback(response.data) : []
}
