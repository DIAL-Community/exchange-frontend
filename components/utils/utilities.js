export const prependUrlWithProtocol = (url) => '//' + url

export const isValidFn = (maybeFunction) =>
  maybeFunction && {}.toString.call(maybeFunction) === '[object Function]'

export const isDebugLoggingEnabled = () => String(process.env.NEXT_PUBLIC_ENABLE_DEBUG).toLowerCase() === 'true'
