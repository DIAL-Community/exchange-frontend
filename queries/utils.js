const MIN_SEARCH_VALUE_CHARACTERS = 2

export const fetchSelectOptions = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}

export const fetchSelectOptionsWithMature = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input,
    mature: true
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}
