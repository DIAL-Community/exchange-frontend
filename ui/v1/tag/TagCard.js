import { useCallback } from 'react'
import { useIntl } from 'react-intl'
import Link from 'next/link'
import parse from 'html-react-parser'
import { FaXmark } from 'react-icons/fa6'
import { DisplayType } from '../utils/constants'
import { isValidFn } from '../utils/utilities'

const TagCard = ({ displayType, index, tag, dismissHandler }) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const displayLargeCard = () =>
    <div className={`px-4 py-6 rounded-lg min-h-[13.5rem] ${index % 2 === 0 && 'bg-dial-violet'}`}>
      <div className='flex flex-col lg:flex-row gap-x-6 gap-y-3'>
        <div className='flex flex-col gap-y-3 max-w-3xl lg:w-10/12'>
          <div className='text-lg font-semibold text-dial-plum'>
            {tag.name}
          </div>
          <div className='line-clamp-4 text-dial-stratos'>
            {tag?.tagDescription && parse(tag?.tagDescription.description)}
          </div>
          <div className='flex gap-x-2 text-dial-stratos'>
            <div className='text-sm'>
              {format('ui.dataset.header')} ({tag.datasets?.length ?? 0})
            </div>
            <div className='border-r border-dial-slate-400' />
            <div className='text-sm'>
              {format('ui.product.header')} ({tag.products?.length ?? 0})
            </div>
            <div className='border-r border-dial-slate-400' />
            <div className='text-sm'>
              {format('ui.project.header')} ({tag.projects?.length ?? 0})
            </div>
            <div className='border-r border-dial-slate-400' />
            <div className='text-sm'>
              {format('ui.useCase.header')} ({tag.useCases?.length ?? 0})
            </div>
          </div>
        </div>
      </div>
    </div>

  const displaySmallCard = () =>
    <div className='rounded-lg bg-gradient-to-r from-tag-bg-light to-tag-bg h-16'>
      <div className='flex flex-row gap-x-3 px-6 h-full'>
        <div className='text-sm font-semibold text-dial-plum my-auto'>
          {tag.name}
        </div>
      </div>
    </div>

  return (
    <div className='relative'>
      <Link href={`/tags/${tag.slug}`}>
        {displayType === DisplayType.LARGE_CARD && displayLargeCard()}
        {displayType === DisplayType.SMALL_CARD && displaySmallCard()}
      </Link>
      { isValidFn(dismissHandler) &&
        <button type='button' className='absolute top-2 right-2'>
          <FaXmark size='1rem' className='text-dial-plum' onClick={dismissHandler} />
        </button>
      }
    </div>
  )
}

export default TagCard
