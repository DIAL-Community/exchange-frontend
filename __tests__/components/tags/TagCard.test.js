import { render } from '../../test-utils'
import CustomMockedProvider from '../../utils/CustomMockedProvider'
import TagCard from '../../../components/tags/TagCard'
import { mockNextAuthUseSession, mockNextUseRouter, statuses } from '../../utils/nextMockImplementation'
import { tag } from './data/TagCard'

mockNextUseRouter()
describe('Unit test for the TagCard component', () => {
  const CARD_TEST_ID = 'tag-card'
  const DELETE_BUTTON_TEST_ID = 'delete-button'

  describe('Should match snapshot -', () => {
    test('user is NOT an admin, displayEditButtons not passed.', () => {
      mockNextAuthUseSession(statuses.AUTHENTICATED, { canEdit: true })
      const { container, getByTestId, queryByTestId } = render(
        <CustomMockedProvider>
          <TagCard
            tag={'Example Tag'}
            listType='list'
          />
        </CustomMockedProvider>
      )
      expect(getByTestId(CARD_TEST_ID)).toHaveTextContent('Example Tag')
      expect(queryByTestId(DELETE_BUTTON_TEST_ID)).not.toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })

    test('user is NOT an admin, displayEditButtons passed.', () => {
      mockNextAuthUseSession(statuses.AUTHENTICATED, { canEdit: false })
      const { container, getByTestId, queryByTestId } = render(
        <CustomMockedProvider>
          <TagCard
            tag={'Example Tag'}
            listType='list'
            displayEditButtons
          />
        </CustomMockedProvider>
      )
      expect(getByTestId(CARD_TEST_ID)).toHaveTextContent('Example Tag')
      expect(queryByTestId(DELETE_BUTTON_TEST_ID)).not.toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })

    test('user is an admin, displayEditButtons not passed.', () => {
      mockNextAuthUseSession(statuses.AUTHENTICATED, { canEdit: true })
      const { container, getByTestId, queryByTestId } = render(
        <CustomMockedProvider>
          <TagCard
            tag={'Example Tag'}
            listType='list'
          />
        </CustomMockedProvider>
      )
      expect(getByTestId(CARD_TEST_ID)).toHaveTextContent('Example Tag')
      expect(queryByTestId(DELETE_BUTTON_TEST_ID)).not.toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })

    test('user is an admin, displayEditButtons passed.', () => {
      mockNextAuthUseSession(statuses.AUTHENTICATED, { canEdit: true })
      const { container, getByTestId } = render(
        <CustomMockedProvider>
          <TagCard
            tag={tag}
            listType='list'
            displayEditButtons
          />
        </CustomMockedProvider>
      )
      expect(getByTestId(CARD_TEST_ID)).toHaveTextContent('Example Tag')
      expect(getByTestId(DELETE_BUTTON_TEST_ID)).toBeInTheDocument()
      expect(container).toMatchSnapshot()
    })
  })
})
