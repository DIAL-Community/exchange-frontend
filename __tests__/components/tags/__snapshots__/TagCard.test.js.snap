// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for the TagCard component Should match snapshot - user is NOT an admin, displayEditButtons not passed. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="border-3 border-transparent text-button-gray"
    data-testid="tag-card"
  >
    <div
      class="border border-dial-gray card-drop-shadow"
    >
      <div
        class="flex gap-3 my-5 px-4"
      >
        <div
          class="card-title text-button-gray"
        >
          <span>
            Example Tag
          </span>
        </div>
        <div
          class="line-clamp-1"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Unit test for the TagCard component Should match snapshot - user is NOT an admin, displayEditButtons passed. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="border-3 border-transparent text-button-gray"
    data-testid="tag-card"
  >
    <div
      class="border border-dial-gray card-drop-shadow"
    >
      <div
        class="flex gap-3 my-5 px-4"
      >
        <div
          class="card-title text-button-gray w-2/6"
        >
          <span>
            Example Tag
          </span>
        </div>
        <div
          class="line-clamp-1"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Unit test for the TagCard component Should match snapshot - user is an admin, displayEditButtons not passed. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="border-3 border-transparent text-button-gray"
    data-testid="tag-card"
  >
    <div
      class="border border-dial-gray card-drop-shadow"
    >
      <div
        class="flex gap-3 my-5 px-4"
      >
        <div
          class="card-title text-button-gray"
        >
          <span>
            Example Tag
          </span>
        </div>
        <div
          class="line-clamp-1"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Unit test for the TagCard component Should match snapshot - user is an admin, displayEditButtons passed. 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="border-3 border-transparent text-button-gray"
    data-testid="tag-card"
  >
    <div
      class="border border-dial-gray card-drop-shadow"
    >
      <div
        class="flex gap-3 my-5 px-4"
      >
        <div
          class="card-title text-button-gray w-2/6"
        >
          <span>
            Example Tag
          </span>
        </div>
        <div
          class="line-clamp-1"
        />
        <div
          class="flex flex-row gap-3 ml-auto"
        >
          <button
            class="flex items-center gap-x-1.5 bg-dial-iris-blue px-2 py-1 rounded-md text-white"
            data-testid="edit-button"
            type="button"
          >
            <img
              alt="Edit"
              height="12px"
              src="/icons/edit.svg"
              width="12px"
            />
            <span
              class="text-sm"
            >
              Edit
            </span>
          </button>
          <button
            class="delete-button gap-x-1.5"
            data-testid="delete-button"
            type="button"
          >
            <svg
              class="text-sm"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5.5 5.5A.5.5 0 0 1 6 6v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5zm2.5 0a.5.5 0 0 1 .5.5v6a.5.5 0 0 1-1 0V6a.5.5 0 0 1 .5-.5zm3 .5a.5.5 0 0 0-1 0v6a.5.5 0 0 0 1 0V6z"
              />
              <path
                d="M14.5 3a1 1 0 0 1-1 1H13v9a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V4h-.5a1 1 0 0 1-1-1V2a1 1 0 0 1 1-1H6a1 1 0 0 1 1-1h2a1 1 0 0 1 1 1h3.5a1 1 0 0 1 1 1v1zM4.118 4 4 4.059V13a1 1 0 0 0 1 1h6a1 1 0 0 0 1-1V4.059L11.882 4H4.118zM2.5 3V2h11v1h-11z"
                fill-rule="evenodd"
              />
            </svg>
            <span
              class="text-sm"
            >
              Delete
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
